modal/app.py

```diff
@@ -258,6 +262,44 @@ def description(self) -> Optional[str]:
         """The App's `name`, if available, or a fallback descriptive identifier."""
         return self._description
 
+    @staticmethod
+    def from_name(name: str, environment_name: str = "", create_if_missing: bool = False) -> "_App":
+        """Look up an App with a given name, creating a new App if necessary.
+
+        Note that Apps created through this method will be in a deployed state,
+        but they will not have any associated Functions or Classes.
+
+        This method is mainly useful for creating an App to associate with a Sandbox:
+
+        ```python
+        app = modal.App.from_name("my-app", create_if_missing=True)
+        app.hydrate()  # Pre-hydrate to accelerate Sandbox creation
+        modal.Sandbox.create("echo", "hi", app=app)
+        ```
+
+        When building an App that will be populated with Functions and Classes, you should
+        call the main `App()` constructor directly.
+        """
+
+        # TODO We need to enforce rules about App names, but we previously weren't in App.lookup
+        # So we probably need to add that gradually (i.e. as a warning first) to avoid breaking people?
+        async def _load(self: _App, client: _Client):
+            request = api_pb2.AppGetOrCreateRequest(
+                app_name=name,
+                environment_name=_get_environment_name(environment_name),
+                object_creation_type=(api_pb2.OBJECT_CREATION_TYPE_CREATE_IF_MISSING if create_if_missing else None),
+            )
+            response = await retry_transient_errors(client.stub.AppGetOrCreate, request)
+            self._app_id = response.app_id
+            self._client = client
+            # TODO should we also be calling AppGetLayout and populating the running_app with it?
+            self._running_app = RunningApp(response.app_id, interactive=False)
+
+        app = _App(name)
+        app._load = _load
```

Comment:
super minor but this could maybe just be a constructor argument

Reply:
I’m fairly averse to “private parameters” in public interfaces: best case, you confuse users or at least present them with information they don’t need to think about, worse case they try to use it and end up in trouble somehow.

Reply:
I’d really prefer to just merge `__init__` and .from_name` but couldn’t see a simple way to do that.